llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp

```diff
@@ -946,8 +946,8 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
       // Check that the instruction return type is vectorizable.
       // We can't vectorize casts from vector type to scalar type.
       // Also, we can't vectorize extractelement instructions.
-      if ((!VectorType::isValidElementType(I.getType()) &&
-           !I.getType()->isVoidTy()) ||
+      Type *InstTy = I.getType();
+      if (!(InstTy->isVoidTy() || canWidenType(InstTy)) ||
```

Comment:
The possible problem I see here is that we're assuming that we can vectorise any arbitrary operation that returns a structure, which in practice may not be true because this patch is currently only testing the case of call instructions. @MacDue have you tested this patch to see what happens on IR like this?

```
  %load = load { i32, i32 }, ptr %gep.A, align 4
```

to see how the vectoriser handles it with your patch? I suspect that if we choose to go with this version of your patch we probably want to restrict struct returns to calls for now.

Reply:
I've added two extra tests with `load` and `store` instructions, and updated the check in `LoopVectorizationLegality.cpp` to only allow widening calls that return structs, where all users are `extractvalue`. 

This is all we need for this initial patch (we could maybe generalize this more in future patches).